src/backend/lib/documentHandler.ts

```diff
@@ -0,0 +1,106 @@
+import type { FastifyReply, FastifyRequest } from 'fastify';
+import type { DocumentStore } from './document_stores/base.js';
+import { createKey } from './keyGenerator.js';
+import type { DocumentHandlerConfig, FastifyRequestGeneric } from './types.js';
+
+/**
+ * Handles document operations such as retrieving, creating, and handling raw versions of documents.
+ */
+export class DocumentHandler {
+	#keyLength: number;
+	#store: DocumentStore;
+
+	/**
+	 * Creates an instance of DocumentHandler.
+	 * @param config The configuration object for DocumentHandler.
+	 */
+	public constructor(config: DocumentHandlerConfig) {
+		this.#keyLength = config.keyLength;
+		this.#store = config.store;
+	}
+
+	/**
+	 * Handle retrieving a document
+	 * @param request The incoming request
+	 * @param reply The outgoing reply
+	 */
+	public async handleGet(request: FastifyRequest<FastifyRequestGeneric>, reply: FastifyReply) {
+		const key = request.params.id.split('.')[0];
+
+		const result = await this.#store.get(key);
+
+		if (result) {
+			return reply //
+				.type('application/json')
+				.code(200)
+				.send(this.isHeadRequest(request) ? undefined : { data: result, key });
+		}
+
+		return reply //
+			.code(404)
+			.send(this.isHeadRequest(request) ? undefined : { message: 'Document not found.' });
+	}
+
+	/**
+	 * Handle retrieving the raw version of a document
+	 * @param request The incoming request
+	 * @param reply The outgoing reply
+	 */
+	public async handleRawGet(request: FastifyRequest<FastifyRequestGeneric>, reply: FastifyReply) {
+		const key = request.params.id.split('.')[0];
+
+		const result = await this.#store.get(key);
+
+		if (result) {
+			return reply //
+				.type('text/plain; charset=UTF-8')
+				.code(200)
+				.raw.end(this.isHeadRequest(request) ? undefined : result);
+		}
+
+		return reply //
+			.code(404)
+			.send(this.isHeadRequest(request) ? undefined : { message: 'Document not found.' });
```

Comment:
The 404 response is `application/json`.

Reply:
Pretty sure fastify sets the content type automatically but I can set it explicitly I guess. Or do you mean that it shouldn't be and that there shouldnt be a custom message?

Reply:
I mean, if fastify sets them implicitly, then you can delete the explicit sets

Reply:
![ishare-1710452700](https://github.com/skyra-project/haste-server/assets/4019718/52f4d9e3-a2b5-4317-a672-83083abaa56e)


Set implicitly